packages/opentelemetry-scope-zone-peer-dep/test/ZoneScopeManager.test.ts

```diff
@@ -162,16 +162,140 @@ describe('ZoneScopeManager', () => {
-    it('should call the function with previously assigned scope', () => {
+  describe.only(".withAsync()", () => {
+    it("should run the callback (null as target)", async done => {
+      await scopeManager.withAsync(null, async () => done());
+    });
+
+    it("should run the callback (object as target)", async done => {
+      const test = { a: 1 };
+      await scopeManager.withAsync(test, async () => {
+        assert.strictEqual(scopeManager.active(), test, "should have scope");
+        done();
+      });
+    });
+
+    it("should run the callback (when disabled)", async done => {
+      scopeManager.disable();
+      await scopeManager.withAsync(null, async () => {
+        scopeManager.enable();
+        return done();
+      });
+    });
+
+    it("should rethrow errors", done => {
+      scopeManager
+        .withAsync(null, async () => {
+          throw new Error("This should be rethrown");
+        })
+        .then(
+          () => {
+            assert.equal("this code", "should never be executed");
+          },
+          () => {
+            done();
+          }
+        );
+    });
+
+    it.only("should finally restore an old scope, including the async task", async () => {
+      const scope1 = "scope1";
+      const scope2 = "scope2";
+      scope2;
+      const scope3 = "scope3";
+      scope3;
+
+      await scopeManager.withAsync(scope1, async () => {
+        assert.strictEqual(scopeManager.active(), "scope1");
+        await scopeManager.withAsync(scope2, async () => {
+          assert.strictEqual(scopeManager.active(), "scope2");
+          await scopeManager.withAsync(scope3, async () => {
+            assert.strictEqual(scopeManager.active(), "scope3");
+          });
+          //   assert.strictEqual(scopeManager.active(), "scope2");
+        });
+        // assert.strictEqual(scopeManager.active(), "scope1");
+        setTimeout(() => {
+          // assert.strictEqual(scopeManager.active(), "scope1");
+        }, 500);
+        clock.tick(500);
+      });
+      assert.strictEqual(scopeManager.active(), window);
+    });
+
+    it("should finally restore an old scope when scope is an object, including the async task", done => {
+      const scope1 = { a: 1 };
+      const scope2 = { a: 2 };
+      const scope3 = { a: 3 };
+      scopeManager.with(scope1, () => {
+        assert.strictEqual(scopeManager.active(), scope1);
+        scopeManager.with(scope2, () => {
+          assert.strictEqual(scopeManager.active(), scope2);
+          scopeManager.with(scope3, () => {
+            assert.strictEqual(scopeManager.active(), scope3);
+          });
+          assert.strictEqual(scopeManager.active(), scope2);
+        });
+        assert.strictEqual(scopeManager.active(), scope1);
+        setTimeout(() => {
+          assert.strictEqual(scopeManager.active(), scope1);
+          done();
+        }, 500);
+        clock.tick(500);
+      });
+      assert.strictEqual(scopeManager.active(), window);
+    });
```

Comment:
ok so I duplicated these tests across from the non async version of the function. I figure we want them to offer exactly the same functionality, just addressing the bug I pointed out in #752 . 

Unfortunately, all the scopes appear to be closing down in the async version as you complete the inner function calls.

So by the time you get to line 245, all the scopes have been closed and the active scope is `window`